File: March 2023/Day 15/Special Palindrome Substrings/special-palindrome-substrings.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
    
	public static void main (String[] args)throws IOException {
	
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());
		while(t-->0)
		{
		    String str = br.readLine();
		    String s1 = str.split(" ")[0];
		    String s2 = str.split(" ")[1];
		    
		    Solution obj = new Solution();
		    int ans = obj.specialPalindrome(s1,s2);
		    System.out.println(ans);
		    
		    
		    
		}
	}
}
// } Driver Code Ends


//User function Template for Java


class Solution{
    public  static int specialPalindrome(String a, String b){
       //Code Here
       int ans=Integer.MAX_VALUE;
       for(int i=0;i<=a.length()-b.length();i++) {
           int val = f(a,b,i,i+b.length()-1);
           if(val!=-1) ans=Math.min(ans,val);
       }
       if(ans==Integer.MAX_VALUE) ans=-1;
       return ans;
    }
    
    public static int f(String a,String b,int strt,int end) {
        int ans=0;
        for(int i=0,j=a.length()-1;i<=j;i++,j--) {
            if(i>=strt && i<=end) {
                if(j>=strt && j<=end) {
                    if(b.charAt(i-strt)!=b.charAt(j-strt)) return -1;
                    if(a.charAt(i)!=b.charAt(i-strt)) ans++;
                    if(a.charAt(j)!=b.charAt(j-strt) && i!=j) ans++;
                } else {
                    // System.out.println(strt-i);
                    if(a.charAt(i)!=b.charAt(i-strt)) ans++;
                    if(b.charAt(i-strt)!=a.charAt(j) && i!=j) ans++;
                }
            } else {
                if(j>=strt && j<=end) {
                    if(a.charAt(j)!=b.charAt(j-strt)) ans++;
                    if(b.charAt(j-strt)!=a.charAt(i) && i!=j) ans++;
                } else {
                    if(a.charAt(i)!=a.charAt(j)) ans++;
                }
            }
        }
        return ans;
    }
}

File: March 2023/Day 15/Special Palindrome Substrings/special-palindrome-substrings.py
#User function Template for python3
import sys
def f(a,b,strt,end) :
    ans=0;
    i=0
    j=len(a)-1
    while i<=j :
        if i>=strt and i<=end :
            if j>=strt and j<=end :
                if b[i-strt]!=b[j-strt] : return -1;
                if a[i]!=b[i-strt] : ans+=1
                if a[j]!=b[j-strt] and i!=j : ans+=1
            else :
                if a[i]!=b[i-strt] : ans+=1
                if b[i-strt]!=a[j] and i!=j : ans+=1
        else :
            if j>=strt and j<=end :
                if a[j]!=b[j-strt] : ans+=1
                if b[j-strt]!=a[i] and i!=j : ans+=1
            else :
                if a[i]!=a[j] : ans+=1
        i+=1
        j-=1
    return ans;
class Solution():
    def specialPalindrome(self,s1, s2):
        #your code goes here
        ans=sys.maxsize
        for i in range(0,len(s1)-len(s2)+1) :
            val = f(s1,s2,i,i+len(s2)-1)
            if val!=-1 : ans=min(ans,val)
            
        if ans==sys.maxsize : ans=-1
        return ans;

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == "__main__":
    for _ in range(int(input())):
        s1, s2 = input().split()
        obj = Solution()
        print(obj.specialPalindrome(s1, s2))
# } Driver Code Ends